### DISCUSSION TITLE
Is Checksum verification supported for uploads to OneDrive whilst using Cryptomator?

### DISCUSSION THREAD

**Discussion by GeraldPreston** [iterate-ch/cyberduck#14408] [Q&A]
Hi there,

I know that verifying file integrity via checksums are supported via S3 but will files uploaded to a cryptomator vault on OneDrive also be verified via Checksum?

I am not sure whether OneDrive would be able to generate a checksum for a file within in an ecnrypted vault like that? Does OneDrive see the individual encrypted files or just the whole vault?

I want to make sure the files I upload don't get any errors in transmission. I've just started using CD so my understanding is a little patchy!

Thanks
Matt

**Comment by AliveDevil** [comment]
A Cryptomator vault is a collection of files, that are encrypted in filename and content.
OneDrive probably provides checksums Cyberduck could verify, but what do we do?

crc32, returned-if available
Sha1, returned-if available
Sha256, returned-if available
QuickXor, returned if available

but OneDrive personal definitely does not provide QuickXor,
OneDrive Business and Sharepoint don’t provide Crc32, Sha1 and Sha256.

https://learn.microsoft.com/en-us/graph/api/resources/hashes

It may be hit or miss whether we can verify checksums, or not - and we currently don’t support Microsoft’s „QuickXor“-hash.

**Reply by AliveDevil** [reply]
Another issue we have is this wonderfully worded note:
> In some cases hash values may not be available. If this is the case, the hash values on an item will be updated after the item is downloaded.

Which means they may or may not be available, whatever the case, you need to download the file, to get the hash value reliably.
So upload, then download, then compare.

**Reply by GeraldPreston** [reply]
Thanks for the info - so it's not definite that OneDrive will always return hash info, but if CB is provided hash info, it will use it?

Is there a way to tell within CB whether a file or folder was confirmed via hash or not?

**Reply by AliveDevil** [reply]
If the protocol supports checksums and Verify Checksum on Upload[^vcou] is enabled, then Cyberduck will compare the uploaded checksum, and Warn if the expected checksum isn't returned.
[^vcou]: [Cyberduck Help/Protocols/Amazon S3#Checksums](https://docs.cyberduck.io/protocols/s3/#checksums)

Though bear in mind, the checksum in the Cyberduck Info-window is the checksum of the file as seen by the remote, that is the hash of the encrypted file, rather than the unencrypted file. You'd have to download the file to get the unencrypted file content, and recompute the hash.

**Reply by GeraldPreston** [reply]
Gotcha, that makes sense. Thanks for the clarification. CD is a great tool!
